File: rust/crates/slophammer-cli/src/rust_rules/scope.rs
use super::definitions::{definition, rule_ids};
use crate::config::{Config, RustConfig};
use crate::core::Finding;
use crate::scan::Snapshot;
use globset::{Glob, GlobSet, GlobSetBuilder};

/// Configured scope must account for every production Rust file: each one is
/// either inside a configured paths/targets scope or covered by an exclude
/// (conventional or reasoned). Anything else is a finding, so narrowing scope
/// cannot hide code from checking.
pub fn scope_findings(snapshot: &Snapshot, config: &Config) -> Vec<Finding> {
    let Some(rust) = &config.rust else {
        return Vec::new();
    };
    let scopes = configured_scopes(rust);
    if scopes.is_empty() {
        return Vec::new();
    }
    let uncovered = uncovered_production_dirs(snapshot, config, &scopes);
    if uncovered.is_empty() {
        return Vec::new();
    }
    let template = definition(rule_ids::RUST_SCOPE_INCOMPLETE);
    vec![Finding::with_message(
        template,
        template.path,
        format!("{}: {}", template.message, uncovered.join(", ")),
    )]
}

/// Scope coverage counts for the report: production files the configured
/// scope covers versus all production files. None when no scope is
/// configured.
pub fn scope_counts(snapshot: &Snapshot, config: &Config) -> Option<(usize, usize)> {
    let rust = config.rust.as_ref()?;
    let scopes = configured_scopes(rust);
    if scopes.is_empty() {
        return None;
    }
    let production = production_rust_files(snapshot);
    let scanned = production
        .iter()
        .filter(|path| in_targets(path, &scopes))
        .count();
    Some((scanned, production.len()))
}

/// Mutation targets participate so narrowing mutation scope stays visible.
fn configured_scopes(rust: &RustConfig) -> Vec<String> {
    let mut scopes = rust.targets.clone();
    if let Some(dry) = &rust.dry {
        scopes.extend(dry.paths.iter().cloned());
    }
    if let Some(coverage) = &rust.coverage {
        scopes.extend(coverage.paths.iter().cloned());
    }
    if let Some(mutation) = &rust.mutation {
        scopes.extend(mutation.targets.iter().cloned());
    }
    scopes
}

fn uncovered_production_dirs(
    snapshot: &Snapshot,
    config: &Config,
    scopes: &[String],
) -> Vec<String> {
    let exclude_set = globset(&all_exclude_patterns(config));
    let mut dirs: Vec<String> = production_rust_files(snapshot)
        .into_iter()
        .filter(|path| !in_targets(path, scopes))
        .filter(|path| {
            exclude_set
                .as_ref()
                .is_none_or(|set| !set.is_match(path.as_str()))
        })
        .map(|path| parent_dir(&path))
        .collect();
    dirs.sort();
    dirs.dedup();
    dirs
}

fn all_exclude_patterns(config: &Config) -> Vec<String> {
    let mut patterns = crate::config::rust_exclude(config);
    patterns.extend(crate::config::rust_dry_exclude(config));
    if let Some(coverage) = config.rust.as_ref().and_then(|rust| rust.coverage.as_ref()) {
        patterns.extend(crate::config::exclude_patterns(&coverage.exclude));
    }
    patterns
}

fn production_rust_files(snapshot: &Snapshot) -> Vec<String> {
    snapshot
        .files
        .keys()
        .filter(|path| path.ends_with(".rs"))
        .filter(|path| !conventional_path(path))
        .cloned()
        .collect()
}

/// Path-level form of the conventional non-production list in
/// specs/CONFIG.md.
fn conventional_path(path: &str) -> bool {
    const DIRS: [&str; 12] = [
        "tests",
        "fixtures",
        "templates",
        "testdata",
        "dist",
        "build",
        "coverage",
        "target",
        "node_modules",
        "vendor",
        "scripts",
        "benches",
    ];
    if path.ends_with("_test.rs") || path.contains("generated") {
        return true;
    }
    path.split('/').any(|segment| DIRS.contains(&segment))
}

fn parent_dir(path: &str) -> String {
    match path.rsplit_once('/') {
        Some((dir, _)) => dir.to_owned(),
        None => ".".to_owned(),
    }
}

pub fn rust_files(snapshot: &Snapshot, paths: &[String], excludes: &[String]) -> Vec<String> {
    let exclude_set = globset(excludes);
    snapshot
        .files
        .keys()
        .filter(|path| path.ends_with(".rs"))
        .filter(|path| in_targets(path, paths))
        .filter(|path| !default_excluded(path))
        .filter(|path| {
            exclude_set
                .as_ref()
                .is_none_or(|set| !set.is_match(path.as_str()))
        })
        .cloned()
        .collect()
}

pub fn configured_rust_files(snapshot: &Snapshot, config: &Config) -> Vec<String> {
    rust_files(
        snapshot,
        &crate::config::rust_targets(config),
        &crate::config::rust_exclude(config),
    )
}

pub fn dry_rust_files(snapshot: &Snapshot, config: &Config) -> Vec<String> {
    rust_files(
        snapshot,
        &crate::config::rust_dry_paths(config),
        &crate::config::rust_dry_exclude(config),
    )
}

fn in_targets(path: &str, targets: &[String]) -> bool {
    targets.is_empty()
        || targets.iter().any(|target| {
            let normalized = target.trim_end_matches('/');
            normalized == "." || path == normalized || path.starts_with(&format!("{normalized}/"))
        })
}

fn default_excluded(path: &str) -> bool {
    path.starts_with("fixtures/")
        || path.starts_with("templates/")
        || path.contains("/fixtures/")
        || path.contains("/target/")
        || path.contains("/node_modules/")
        || path.ends_with("_test.rs")
}

fn globset(patterns: &[String]) -> Option<GlobSet> {
    if patterns.is_empty() {
        return None;
    }
    let mut builder = GlobSetBuilder::new();
    for pattern in patterns {
        if let Ok(glob) = Glob::new(pattern) {
            builder.add(glob);
        }
    }
    builder.build().ok()
}

File: python/src/slophammer_py/rules/scope.py
"""Scope completeness: configured scope must account for every production
Python file, so narrowing scope cannot hide code from checking. Coverage
paths, DRY paths, and mutation targets all participate in the scope union.
"""

from __future__ import annotations

import fnmatch

from ..config import Config, ExcludeEntry, PythonConfig
from ..core import Finding, ScopeCoverage
from ..repo import Snapshot
from .definitions import Definition

CONVENTIONAL_DIRS = {
    "tests",
    "test",
    "fixtures",
    "templates",
    "testdata",
    "dist",
    "build",
    "coverage",
    "target",
    "node_modules",
    "vendor",
    "scripts",
    "migrations",
}


def scope_findings(definition: Definition, snapshot: Snapshot, config: Config) -> list[Finding]:
    scopes = configured_scopes(config.python)
    if not scopes or config.python is None:
        return []
    uncovered = uncovered_production_dirs(snapshot, config.python, scopes)
    if not uncovered:
        return []
    return [
        Finding(
            rule_id=definition.id,
            severity=definition.severity,
            path=definition.path,
            message=f"{definition.message}: {', '.join(uncovered)}",
        )
    ]


def scope_counts(snapshot: Snapshot, config: Config) -> ScopeCoverage | None:
    scopes = configured_scopes(config.python)
    if not scopes:
        return None
    production = production_python_files(snapshot)
    scanned = sum(1 for path in production if in_targets(path, scopes))
    return ScopeCoverage(scanned=scanned, production_files=len(production))


def configured_scopes(python: PythonConfig | None) -> list[str]:
    if python is None:
        return []
    scopes: list[str] = []
    if python.coverage is not None:
        scopes.extend(python.coverage.paths)
    if python.dry is not None:
        scopes.extend(python.dry.paths)
    if python.mutation is not None:
        scopes.extend(python.mutation.targets)
    return scopes


def uncovered_production_dirs(
    snapshot: Snapshot, python: PythonConfig, scopes: list[str]
) -> list[str]:
    patterns = all_exclude_patterns(python)
    directories = {
        parent_dir(path)
        for path in production_python_files(snapshot)
        if not in_targets(path, scopes) and not excluded(path, patterns)
    }
    return sorted(directories)


def all_exclude_patterns(python: PythonConfig) -> list[str]:
    sections: list[list[ExcludeEntry]] = []
    if python.coverage is not None:
        sections.append(python.coverage.exclude)
    if python.dry is not None:
        sections.append(python.dry.exclude)
    if python.mutation is not None:
        sections.append(python.mutation.exclude)
    return [entry.pattern for entries in sections for entry in entries]


def production_python_files(snapshot: Snapshot) -> list[str]:
    return [path for path in snapshot.files if path.endswith(".py") and not conventional_path(path)]


def conventional_path(path: str) -> bool:
    base = path.rsplit("/", 1)[-1]
    if base.startswith("test_") or base.endswith("_test.py") or base == "conftest.py":
        return True
    if "generated" in path:
        return True
    return any(segment in CONVENTIONAL_DIRS for segment in path.split("/")[:-1])


def parent_dir(path: str) -> str:
    return path.rsplit("/", 1)[0] if "/" in path else "."


def in_targets(path: str, targets: list[str]) -> bool:
    if not targets:
        return True
    for target in targets:
        normalized = target.rstrip("/")
        if normalized in (".", "") or path == normalized or path.startswith(normalized + "/"):
            return True
    return False


def excluded(path: str, patterns: list[str]) -> bool:
    return any(matches_pattern(path, pattern) for pattern in patterns)


def matches_pattern(path: str, pattern: str) -> bool:
    if fnmatch.fnmatch(path, pattern):
        return True
    # `dir/**` covers everything under dir, including dir itself.
    if pattern.endswith("/**"):
        return path.startswith(pattern[:-3] + "/") or fnmatch.fnmatch(path, pattern[:-3])
    return False
